Opcodes/emugens/emugens.c

```diff
@@ -18,255 +18,527 @@
-}
-
-#define INTERP_M(X, X0, X1, Y0, Y1) (((X)-(X0))/((X1)-(X0)) * ((Y1)-(Y0)) + (Y0))
+   ky  bpf  kx, kx0, ky0, kx1, ky1, ...
+   kY[] bpf kX[], kx0, ky0, kx1, ky1, ...
 
-inline MYFLT interpol_m(MYFLT x, MYFLT x0, MYFLT x1, MYFLT y0, MYFLT y1) {
-    return (x-x0)/(x1-x0) * (y1-y0) + y0;
-}
+ */
 
+#define BPF_MAXPOINTS 256
 
 typedef struct {
-  OPDS    h;
-  MYFLT *r, *x, *x0, *y0, *x1, *y1, *x2, *y2;
-} BPF3;
-
+    OPDS h;
+    MYFLT *r, *x, *data[BPF_MAXPOINTS];
+} BPFX;
 
-static int32_t bpf3(CSOUND *csound, BPF3 *p) {
-    IGN(csound);
+static int32_t bpfx(CSOUND *csound, BPFX *p) {
     MYFLT x = *p->x;
-    MYFLT n, m;
-    if(x<*p->x1) {
-      m = *p->x0; n = *p->y0;
-      *p->r = INTERP_L(x, m, *p->x1, n, *p->y1);
-    } else {
-      m = *p->x1;
-      n = *p->y1;
-      *p->r = INTERP_R(x, m, *p->x2, n, *p->y2);
+    MYFLT **data = p->data;
+    int32_t datalen = p->INOCOUNT - 1;
+    if(datalen % 2) {
+        return INITERR(Str("bpf: data length should be even (pairs of x, y)"));
     }
-    return OK;
+    if(datalen >= BPF_MAXPOINTS) {
+        return INITERR(Str("bpf: too many pargs (max=256)"));
+    }
+    int32_t i;
+    MYFLT x0, x1, y0, y1;
+    x0 = *data[0];
+    y0 = *data[1];
+
+    if (x <= x0) {
+        *p->r = y0;
+        return OK;
+    }
+    if (x>=*data[datalen-2]) {
+        *p->r = *data[datalen-1];
+        return OK;
+    }
+    for(i=2; i<datalen; i+=2) {
+        x1 = *data[i];
+        y1 = *data[i+1];
+        if( x <= x1 ) {
+            *p->r = (x-x0)/(x1-x0)*(y1-y0)+y0;
+            return OK;
+        }
+        x0 = x1;
+        y0 = y1;
+    }
+    return NOTOK;
 }
 
-typedef struct {
-  OPDS    h;
-  MYFLT *r, *x, *x0, *y0, *x1, *y1, *x2, *y2, *x3, *y3;
-} BPF4;
-
-
-static int32_t bpf4(CSOUND *csound, BPF4 *p) {
-    IGN(csound);
+static int32_t bpfxcos(CSOUND *csound, BPFX *p) {
     MYFLT x = *p->x;
-    MYFLT m, n;
-    if(x < (*p->x1)) {
-      m = *p->x0; n = *p->y0;
-      *p->r = INTERP_L(x, m, *p->x1, n, *p->y1);
-    } else if (x < (*p->x2)) {
-      m = *p->x1; n = *p->y1;
-      *p->r = INTERP_M(x, m, *p->x2, n, *p->y2);
-    }  else {
-      m = *p->x2; n = *p->y2;
-      *p->r = INTERP_R(x, m, *p->x3, n, *p->y3);
+    MYFLT **data = p->data;
+    int32_t datalen = p->INOCOUNT - 1;
+    if(datalen % 2)
+        return INITERR(Str("bpf: data length should be even (pairs of x, y)"));
+    if(datalen >= BPF_MAXPOINTS)
+        return INITERR(Str("bpf: too many pargs (max=256)"));
+    int32_t i;
+    MYFLT x0, x1, y0, y1, dx;
+    x0 = *data[0];
+    y0 = *data[1];
+
+    if (x <= x0) {
+        *p->r = y0;
+        return OK;
     }
-    return OK;
+    if (x>=*data[datalen-2]) {
+        *p->r = *data[datalen-1];
+        return OK;
+    }
+    for(i=2; i<datalen; i+=2) {
+        x1 = *data[i];
+        y1 = *data[i+1];
+        if( x <= x1 ) {
+            dx = ((x-x0) / (x1-x0)) * M_PI + M_PI;
+            *p->r = y0 + ((y1 - y0) * (1 + cos(dx)) / 2.0);
```

Comment:
You need to `#include <math.h>` for `M_PI`, `cos` and other math functions. I think it;s better to be explicit about the dependencies rather than rely on it being included indirectly (this might fix the appveyor issue).

Reply:
ok, fixed both issues

Reply:
Surely people should be using PI defined in csoundCore.h with the PI_F variant